gen/nvml/generateapi.go

```diff
@@ -0,0 +1,420 @@
+# limitations under the License.
+**/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"slices"
+	"sort"
+	"strings"
+	"unicode"
+)
+
+type GeneratableInterfacePoperties struct {
+	Type                      string
+	Interface                 string
+	Exclude                   []string
+	PackageMethodsAliasedFrom string
+}
+
+var GeneratableInterfaces = []GeneratableInterfacePoperties{
+	{
+		Type:                      "library",
+		Interface:                 "Interface",
+		Exclude:                   []string{"Lookup"},
+		PackageMethodsAliasedFrom: "libnvml",
+	},
+	{
+		Type:      "nvmlDevice",
+		Interface: "Device",
+	},
+	{
+		Type:      "nvmlGpuInstance",
+		Interface: "GpuInstance",
+	},
+	{
+		Type:      "nvmlComputeInstance",
+		Interface: "ComputeInstance",
+	},
+	{
+		Type:      "nvmlEventSet",
+		Interface: "EventSet",
+	},
+	{
+		Type:      "nvmlGpmSample",
+		Interface: "GpmSample",
+	},
+	{
+		Type:      "nvmlUnit",
+		Interface: "Unit",
+	},
+	{
+		Type:      "nvmlVgpuInstance",
+		Interface: "VgpuInstance",
+	},
+	{
+		Type:      "nvmlVgpuTypeId",
+		Interface: "VgpuTypeId",
+	},
+}
+
+func main() {
+	sourceDir := flag.String("sourceDir", "", "Path to the source directory for all go files")
+	output := flag.String("output", "", "Path to the output file (default: stdout)")
+	flag.Parse()
+
+	// Check if required flags are provided
+	if *sourceDir == "" {
+		flag.Usage()
+		return
+	}
+
+	writer, closer, err := getWriter(*output)
+	if err != nil {
+		fmt.Printf("Error: %v", err)
+		return
+	}
+	defer closer()
+
+	header, err := generateHeader()
+	if err != nil {
+		fmt.Printf("Error: %v", err)
+		return
+	}
+	fmt.Fprintf(writer, header)
+
+	for i, p := range GeneratableInterfaces {
+		if p.PackageMethodsAliasedFrom != "" {
+			comment, err := generatePackageMethodsComment(p)
+			if err != nil {
+				fmt.Printf("Error: %v", err)
+				return
+			}
+			fmt.Fprintf(writer, comment)
+
+			output, err := generatePackageMethods(*sourceDir, p)
+			if err != nil {
+				fmt.Printf("Error: %v", err)
+				return
+			}
+			fmt.Fprintf(writer, "%s\n", output)
+		}
+
+		comment, err := generateInterfaceComment(p)
+		if err != nil {
+			fmt.Printf("Error: %v", err)
+			return
+		}
+		fmt.Fprintf(writer, comment)
+
+		output, err := generateInterface(*sourceDir, p)
+		if err != nil {
+			fmt.Printf("Error: %v", err)
+			return
+		}
+		fmt.Fprintf(writer, output)
+
+		if i < (len(GeneratableInterfaces) - 1) {
+			fmt.Fprintf(writer, "\n")
+		}
+	}
+}
+
+func getWriter(outputFile string) (io.Writer, func() error, error) {
+	if outputFile == "" {
+		return os.Stdout, func() error { return nil }, nil
+	}
+
+	file, err := os.Create(outputFile)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return file, file.Close, nil
+}
+
+func generateHeader() (string, error) {
+	lines := []string{
+		"/**",
+		"# Copyright 2024 NVIDIA CORPORATION",
```

Comment:
As a follow up we may want to read a `boilerplate.txt` from disk instead so as to not have to update the date.